### PR TITLE
Allow deletion of replaced nodes

### PR DESCRIPTION
**Pull request by mdrahman-suse**
#### Proposed Changes ####

Address: https://github.com/rancher/distros-test-framework/issues/164

No
#### Types of Changes ####

Fix

#### Testing ####

Checklist:
1. If your PR changes anything on or related to Jenkins, run it pointing to your branch to make sure it's okay.

2. Verify code lint; we should not have errors.

3. Update the documentation if needed.

4. Update makefile and docker run if adding new tests.

5. Run your tests at least 4 times with all configurations needed and possible.

6. If needed test with different os types.

### CODE REVIEW

**Review by endawkins — DISMISSED** [review]
LGTM

**Review by VestigeJ — DISMISSED** [review]

**Review by aganesh-suse — APPROVED** [review]
LGTM

File: pkg/testcase/upgradenodereplacement.go [hunk 1]
```diff
@@ -598,3 +598,25 @@ func buildRke2Cmd(
 
 	return cmd, nil
 }
+
+// DeleteReplacedNodes deletes the new nodes created as replacements.
+func DeleteReplacedNodes(cluster *shared.Cluster) {
```
Comment (fmoral2) [inline pkg/testcase/upgradenodereplacement.go:603]: this really should not live here as a test, you are narrow it down too much

this should live on cluster config, and we have destroy cluster there, we should have a function `deleteAwsInstances` and call it wherever we want on suites that might need ,  not being specific for node replace

**Review by fmoral2 — CHANGES_REQUESTED** [review]
delete func and flag

File: pkg/testcase/upgradenodereplacement.go [hunk 2]
```diff
@@ -598,3 +598,25 @@ func buildRke2Cmd(
...
+	externalIPs := shared.FetchNodeExternalIPs()
```
Comment (fmoral2) [inline pkg/testcase/upgradenodereplacement.go:604]: you dont need this func since you have the cluster data 

```
   var ips []string
    ips = append(cluster.ServerIPs, ips...)
    ips = append(cluster.AgentIPs, ips...)


for _, ip : range ips ....

```

File: pkg/testcase/upgradenodereplacement.go [hunk 3]
```diff
@@ -598,3 +598,25 @@ func buildRke2Cmd(
...
+		wg.Add(1)
+		go func(ip string) {
+			defer wg.Done()
+			nodeDelErr := awsClient.DeleteInstance(ip)
+			if nodeDelErr != nil {
+				shared.LogLevel("error", "on deleting node with ip: %v, got error %w", ip, nodeDelErr)
+				return
+			}
+		}(ip)
+		wg.Wait()
```
Comment (fmoral2) [inline pkg/testcase/upgradenodereplacement.go:620]: the wait should be outside for loop, this way you are actually doing nothing

File: pkg/testcase/upgradenodereplacement.go [hunk 4]
```diff
@@ -598,3 +598,25 @@ func buildRke2Cmd(
 
 	return cmd, nil
 }
+
+// DeleteReplacedNodes deletes the new nodes created as replacements.
+func DeleteReplacedNodes(cluster *shared.Cluster) {
+	externalIPs := shared.FetchNodeExternalIPs()
+	awsClient, err := aws.AddAWSClient(cluster)
+	Expect(err).NotTo(HaveOccurred(), "error creating aws client: %s", err)
```
Comment (fmoral2) [inline pkg/testcase/upgradenodereplacement.go:606]: to print error  %v or %w

File: pkg/testcase/upgradenodereplacement.go (hunk 1 above)
Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go:603]: So this cannot be added to shared/clusterconfig.go due to circular dependency with aws.go... any other suggestion where this can live?

File: pkg/testcase/upgradenodereplacement.go (hunk 3 above)
Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go:620]: FYI I copied the code from here which has similar wait inside the for loop... so this needs to be updated as well then? https://github.com/rancher/distros-test-framework/blob/main/entrypoint/rebootinstances/rebootinstances_suite_test.go#L74

**Comment by mdrahman-suse** [comment]
Updated please re-review

File: pkg/testcase/upgradenodereplacement.go (hunk 1 above)
Comment (fmoral2) [reply pkg/testcase/upgradenodereplacement.go]: yeah makes sense. since this is actual aws implementation, should live in aws.go? 

then on suite 
```

var _ = AfterSuite(func() {
	if customflag.ServiceFlag.Destroy {
		aws.DeleteAwsNodes(cluster)
	}
})

```

File: pkg/testcase/upgradenodereplacement.go (hunk 3 above)
Comment (fmoral2) [reply pkg/testcase/upgradenodereplacement.go]: yes please, sorry about that , doing that way we are actually not using the async benefits

Comment (fmoral2) [reply pkg/testcase/upgradenodereplacement.go]: if you dont mind updating, tks

File: pkg/testcase/upgradenodereplacement.go (hunk 1 above)
Comment (fmoral2) [reply pkg/testcase/upgradenodereplacement.go]: also on suite you will need to a var for aws here 
```
var (
	kubeconfig string
	flags      *customflag.FlagConfig
	cluster    *shared.Cluster
)
```

to be able to reuse it on delete function and also pass it to `testcase.TestUpgradeReplaceNode(cluster,aws,  flags)`

File: pkg/testcase/upgradenodereplacement.go (hunk 3 above)
Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go]: np will update

File: pkg/testcase/upgradenodereplacement.go (hunk 2 above)
Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go]: this approach does not work as it tries to find the old ips, which got deleted as part of tf destroy...

File: pkg/testcase/upgradenodereplacement.go (hunk 1 above)
Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go]: So we already have `DeleteInstance` in aws.go, which I have used in this func so I don't think it should live there as it deals with cluster stuff... one of the reason I wanted to have a `helper`/`support` dir that will live along with or close to the testcases

File: entrypoint/upgradecluster/upgradenodereplacement_test.go [hunk 5]
```diff
@@ -80,6 +79,12 @@ var _ = Describe("Test:", func() {
 			testcase.TestServiceLoadBalancer(false, true)
 		})
 	}
+
+	if flags.Destroy {
```
[the remark this answers is not in the document]
Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go]: Actually adding it on the test makes sense as it is applicable to only this test, so keeping it as is

File: entrypoint/restartservice/restartservice_test.go [hunk 6]
```diff
@@ -101,8 +101,8 @@ var _ = Describe("Test:", func() {
 		})
 	}
 
-	It("Displays Cluster details", func() {
-		testcase.TestDisplayClusterDetails()
+	It("Displays Cluster details after tests", func() {
+		testcase.DisplayClusterDetails()
```
Comment (ShylajaDevadiga) [inline entrypoint/restartservice/restartservice_test.go:105]: 1. Is this change related to this PR? If not its unnecessary for the reviewer to chase this change in the PR context.
2. Why do we need DisplayClusterDetails, why not call the underlying methods instead?

Comment (fmoral2) [reply entrypoint/restartservice/restartservice_test.go:105]: actually this should not live here since its not a test, not a specific test helper.

should live on shared and not also be used as a `It` spec, since its not a actual test or feat

File: entrypoint/upgradecluster/upgradenodereplacement_test.go [hunk 7]
```diff
@@ -80,6 +79,10 @@ var _ = Describe("Test:", func() {
 			testcase.TestServiceLoadBalancer(false, true)
 		})
 	}
+
+	It("Deletes nodes created by aws cli", func() {
+		testcase.DeleteAWSInstance(cluster)
```
Comment (fmoral2) [inline entrypoint/upgradecluster/upgradenodereplacement_test.go:84]: should live on shared and not also be used as a It spec, since its not a actual test or feat

File: pkg/testcase/upgradenodereplacement.go (hunk 1 above)
Comment (fmoral2) [reply pkg/testcase/upgradenodereplacement.go]: upgradeerplacement_test.go
```


.....



var _ = AfterAll(func() {
	ips := append(cluster.ServerIPs, cluster.AgentIPs...)
	for _, ip := range ips {
		err := aws.DeleteInstance(ip)
		Expect(err).ToNot(HaveOccurred())
	}
})


```

Comment (fmoral2) [reply pkg/testcase/upgradenodereplacement.go]: suite
```
var (
	kubeconfig string
	flags      *customflag.FlagConfig
	cluster    *shared.Cluster
	aws        *c.Client
)

func TestMain(m *testing.M) {
	flags = &customflag.ServiceFlag
	flag.Var(&flags.InstallMode, "installVersionOrCommit", "Upgrade with version or commit")
	flag.Var(&flags.Channel, "channel", "channel to use on upgrade")
	flag.Var(&flags.Destroy, "destroy", "Destroy cluster after test")
	flag.Var(&flags.SUCUpgradeVersion, "sucUpgradeVersion", "Version for upgrading using SUC")
	flag.Parse()

	_, err := config.AddEnv()
	if err != nil {
		shared.LogLevel("error", "error adding env vars: %w\n", err)
		os.Exit(1)
	}

	kubeconfig = os.Getenv("KUBE_CONFIG")
	if kubeconfig == "" {
		// gets a cluster from terraform.
		cluster = shared.ClusterConfig()
	} else {
		// gets a cluster from kubeconfig.
		cluster = shared.KubeConfigCluster(kubeconfig)
	}

	aws, err = c.AddClient(cluster)
	Expect(err).NotTo(HaveOccurred())

	os.Exit(m.Run())
}
```

File: pkg/testcase/upgradenodereplacement.go (hunk 2 above)
Comment (fmoral2) [reply pkg/testcase/upgradenodereplacement.go]: https://github.com/rancher/distros-test-framework/pull/166#discussion_r1788063328

File: entrypoint/restartservice/restartservice_test.go (hunk 6 above)
Comment (mdrahman-suse) [reply entrypoint/restartservice/restartservice_test.go]: Updated

File: pkg/testcase/upgradenodereplacement.go (hunk 4 above)
Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go]: Updated

**Review by fmoral2 — DISMISSED** [review]

**Comment by mdrahman-suse** [comment]
Updated please review

File: entrypoint/upgradecluster/upgradenodereplacement_test.go [hunk 8]
```diff
@@ -4,15 +4,17 @@ package upgradecluster
...
-	"github.com/rancher/distros-test-framework/pkg/customflag"
+	"github.com/rancher/distros-test-framework/pkg/aws"
 	"github.com/rancher/distros-test-framework/pkg/testcase"
+	"github.com/rancher/distros-test-framework/shared"
 
 	. "github.com/onsi/ginkgo/v2"
 )
 
-var _ = Describe("Test:", func() {
+var _ = Describe("Upgrade Node Replcement Test:", Ordered, func() {
```
Comment (fmoral2) [inline entrypoint/upgradecluster/upgradenodereplacement_test.go:17]: typo on `Replcement`

File: entrypoint/upgradecluster/upgradenodereplacement_test.go [hunk 9]
```diff
@@ -80,8 +82,35 @@ var _ = Describe("Test:", func() {
...
 		})
 	}
+
+	AfterAll(func() {
+		deleteAWSResources()
+	})
+		
 })
 
+func deleteAWSResources() {
```
Comment (fmoral2) [inline entrypoint/upgradecluster/upgradenodereplacement_test.go:92]: can we add that on suite file? just to keep those files as clean as possible, without any logics ?

Comment (fmoral2) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:92]: also on suite if we can create aws client on main and pass it through like we do with cluster,flags etc... just to keep the pattern of initializing dependencies on main.

Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:92]: so I tried doing that, not sure why but it seems to fetch the node external ip from the previous cluster

File: entrypoint/upgradecluster/upgradenodereplacement_test.go (hunk 8 above)
Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go]: Updated

File: scripts/delete_resources.sh [hunk 10]
```diff
@@ -263,13 +263,13 @@ if [ "${RESOURCES}" = "" ]; then
     fi
 
     #Validate path to the tfvars file
-    if [[ ! -f "${$CONFIG_DIR}"/"$PRODUCT_NAME".tfvars ]]; then
+    if [[ ! -f "${CONFIG_DIR}"/"$PRODUCT_NAME".tfvars ]]; then
```
Comment (fmoral2) [inline scripts/delete_resources.sh:266]: tks for seeing that!

File: entrypoint/upgradecluster/upgradenodereplacement_test.go (hunk 9 above)
Comment (fmoral2) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:94]: Also cannot call it on the suite as the suite applies to all the test and this func is needed only for this test

>  agreed

Comment (fmoral2) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:94]: thats really odd, cause the `aws.AddClient` has nothing to do with cluster.IPS stuff.

Comment (fmoral2) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:94]: but thats ok we can look at that later

**Review by fmoral2 — APPROVED** [review]

File: entrypoint/upgradecluster/upgradenodereplacement_test.go (hunk 9 above)
Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:94]: Yeah I was confused as well as to why it was not working like that

Comment (fmoral2) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:94]: sorry ....really not trying to be picky , but this is like we are doing something that we fought not do to until today lol.....adding logic on  IT specs entrypoint.....

can we figure it out a way do call use this on the logic place? since you are adding this as a specific helper for node replace, can we add this on `/pkg/testcase/nodereplacement.go` ? 

i know that adding there will be a lit bit of trick , but sure is doable

Comment (fmoral2) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:94]: @mdrahman-suse

Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go:94]: Let me try

Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go]: Updated, please take a look @fmoral2

Comment (fmoral2) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go]: sorry for all . let me try again 


i wanted to mean that it should not live on  `/entrypoint/...`  on a IT spec , since its not a test or nor a feature, and even adding there a just a after something was polluting that file , which in theory should be like a file for " outsiders" to see whats been tested on a feature/flow without needing to code on actual code 
not saying that i like it, just saying its purpose.


since we are treating that ( and i agree ) as a upgrade node replacement helper , as other ones like 
`delErr := deleteRemainServer(serverLeaderIP, awsDependencies)`
`func nodeReplaceServers(`
`serverJoin()`
and etc, this should live on the same file or as closest as possible.


- Adding that as a aws utils does not suite better , because the idea of a client pkg, is kind of  "provide an interface and methods of something, so you can use it wherever its needed" so you probably should be exporting this `DeleteEC2Instances()` as a method, not as a util func, here util func would be util or helpers for the pkg itself, not for  outside , similar to this here 
- https://github.com/rancher/distros-test-framework/pull/177/files#diff-d050ce959548e8086404b5d8d53e4f2ddce5917e83a082106bd34416b306244b

And adding that as a method would not make much sense cause we already have a method exported `DELETEINSTANCE`


Thats why where it would be fitting nicer is on the actual test case , because there you are importing a client to make use of it ( aws) and based on methods provided ( DELETE INSTANCE) you can use it and create a custom helper for your needs `DeleteEC2Instances()`

Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go]: So if we add the func in the same file, when and where should we call the func?

**Comment by mdrahman-suse** [comment]
Updated please review @fmoral2

File: entrypoint/upgradecluster/upgradenodereplacement_test.go [hunk 11]
```diff
@@ -6,13 +6,13 @@ import (
 	"fmt"
 
 	"github.com/rancher/distros-test-framework/pkg/assert"
-	"github.com/rancher/distros-test-framework/pkg/customflag"
 	"github.com/rancher/distros-test-framework/pkg/testcase"
+	"github.com/rancher/distros-test-framework/pkg/testcase/helper"
```
Comment (ShylajaDevadiga) [inline entrypoint/upgradecluster/upgradenodereplacement_test.go:10]: Can we call it helpers instead?

Comment (mdrahman-suse) [reply entrypoint/upgradecluster/upgradenodereplacement_test.go]: Changes it to `support`

**Review by ShylajaDevadiga — APPROVED** [review]

File: pkg/testcase/support/awsutils.go [hunk 12]
```diff
@@ -1,14 +1,16 @@
-package aws
+package support
```
Comment (fmoral2) [inline pkg/testcase/support/awsutils.go:1]: do you mind the file be something like: 
- `aws.go`  or `instances.go`.  `ec2s.go` , `infra.go`. something on this path ?

Comment (fmoral2) [reply pkg/testcase/support/awsutils.go:1]: @mdrahman-suse

Comment (fmoral2) [reply pkg/testcase/support/awsutils.go:1]: we dont need utils in the name, since its already in a support/helper pkg

File: pkg/testcase/upgradenodereplacement.go [hunk 13]
```diff
@@ -27,13 +28,14 @@ func TestUpgradeReplaceNode(cluster *shared.Cluster, flags *customflag.FlagConfi
...
 	resourceName := os.Getenv("resource_name")
+	fmt.Println("resource-name: ", resourceName)
 	awsDependencies, err := aws.AddClient(cluster)
 	Expect(err).NotTo(HaveOccurred(), "error adding aws nodes: %s", err)
 
 	// create server names.
 	var serverNames, instanceServerIds, newExternalServerIps, newPrivateServerIps []string
 	for i := 0; i < len(cluster.ServerIPs); i++ {
-		serverNames = append(serverNames, fmt.Sprintf("%s-distros-qa-server-replace-%d", resourceName, i+1))
+		serverNames = append(serverNames, fmt.Sprintf("%s-%s-server-replace%d", resourceName, resourceTag, i+1))
```
Comment (fmoral2) [inline pkg/testcase/upgradenodereplacement.go:38]: sorry we dont need `resourceTag`

I forgot to remove, since its already set on aws.
https://github.com/rancher/distros-test-framework/blob/4f65fe8b7223db7bd9bfabea68e92165e8ccc47a/pkg/aws/aws.go#L321

Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go:38]: I actually removed it from there and added another tag called `Team: distros-qa`... else it will be like `server-replace1-distros-qa`...

Comment (mdrahman-suse) [reply pkg/testcase/upgradenodereplacement.go]: Done

File: pkg/testcase/support/awsutils.go (hunk 12 above)
Comment (mdrahman-suse) [reply pkg/testcase/support/awsutils.go]: Updated
